### PR TITLE
feature/true-zero-copy-4

### PR DESCRIPTION
**Pull request by rex-schilasky**
**Pull request type**

Please check the type of change your PR introduces:
- [X] Feature

**What is the current behavior?**
eCAL's current zero copy mode is realizing zero copy on the receiving side only. The sending (publishing) side still makes a single memory copy to provide a generic concept for all transport layers (zero copy and none zero copy capable ones).

**What is the new behavior?**
The new approach applies true zero copy for local publish/subscribe connections. It enables the direct modification of the memory file content without any additional copy action (zero copy).

The pictures shows two applications `ecal_sample_latency_snd -z` and `ecal_sample_latency_rec` running on Ubuntu 22.04, exchanging 32MB payloads in 3 µs.

Screenshot from 2023-04-11 17-20-48

**Does this introduce a breaking change?**

- [X] Yes

New `CPublisher` API functions added using the new `eCAL::payload` class.

**Sample Usage**

```cpp
#include <ecal/ecal.h>
#include <ecal/ecal_publisher.h>

// binary payload writer class to allow zero copy memory operations
//
// the `Write`and `Update` calls may operate on the target memory file directly (zero copy mode)
class CBinaryPayloadWriter : public eCAL::CPayloadWriter
{
public:
  CBinaryPayloadWriter(size_t size_) : m_size(size_) {}

  // the provisioned memory is uninitialized ->
  // perform a full write operation
  bool Write(void* buffer_, size_t size_) override
  {
    memset(buffer_, 42, size_);
    return true;
  };

  // the provisioned memory is initialized and contains the data from the last write operation ->
  // perform a partial write operation or just modify a few bytes here
  bool Update(void* buffer_, size_t size_) override
  {
    size_t const write_idx((m_clock % 1024) % size_);
    char   const write_chr(m_clock % 10 + 48);
    static_cast<char*>(buffer_)[write_idx] = write_chr;
    m_clock++;
    return true;
  };

  // provide the size of the required memory (eCAL needs to allocate for you).
  size_t GetSize() override { return m_size; };

private:
  size_t m_size = 0;
  int    m_clock = 0;
};

// main entry
int main(int argc, char** argv)
{
  // initialize eCAL API
  eCAL::Initialize(argc, argv, "zero_copy_snd");

  // create payload
  CBinaryPayloadWriter binary_payload_writer(8 * 1024 * 1024);

  // create publisher
  eCAL::CPublisher pub("zero_copy_pub");

  // enable zero copy mode
  pub.ShmEnableZeroCopy(true);

  // enable handshake mode
  pub.ShmSetAcknowledgeTimeout(50 /* ms */);

  // send updates
  while (eCAL::Ok())
  {
    // send content
    pub.Send(binary_payload_writer);
  }

  // destroy publisher
  pub.Destroy();

  // finalize eCAL API
  eCAL::Finalize();

  return(0);
}
```

### CODE REVIEW

**Review by KerstinKeller — COMMENTED** [review]
I would like to discuss the following points before we merge this PR:

- `CPayload` class: Should we really have two functions  `WriteComplete` and `WritePartial` or just `Write` with an enum indicating how to write? I really don't know why, but I think the flag version would have its advantages. But as I said, not sure why.

- I don't like the "fake"/theoretical numbers. You will never achieve these numbers with Protobuf.. eCAL is great and fast, for *real*  payloads.  We should rather emphasize that!

File: ecal/core/include/ecal/ecal_payload.h [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+#include <cstddef>
+
+namespace eCAL
+{
+  // base payload class to allow zero memory operation
+  // 
+  // the `WriteComplete`and `WritePartial` calls perform 
+  // write/copy operations on the target memory directly
+  class CPayload
```
Comment (KerstinKeller) [inline ecal/core/include/ecal/ecal_payload.h:35]: @danieldube Can you give us feedback on the design of this class?

**Review by danieldube — COMMENTED** [review]
Hi Kerstin, thanks for the ping. I'm missing the whole picture, so I probably made some assumptions. Sorry if they are wrong. 
I just took a look on `CPayload` and left some comments and questions in the pull request. @Kerstin-Keller, hope that helps.

File: ecal/core/include/ecal/ecal_payload.h [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    // the provisioned memory is uninitialized ->
+    // perform a full write operation
+    virtual bool WriteComplete (void* buf_, size_t len_) = 0;
+
+    // the provisioned memory is initialized and contains the data from the last write operation ->
+    // perform a partial write operation or just modify a few bytes here
+    //
+    // by default this operation will just call `WriteComplete`
+    virtual bool WritePartial  (void* buf_, size_t len_) { return WriteComplete(buf_, len_); };
```
[the remark this answers is not in the document]
Comment (rex-schilasky) [reply ecal/core/include/ecal/ecal_payload.h:55]: I will rename all for better readability.

File: ecal/core/include/ecal/ecal_payload.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (rex-schilasky) [reply ecal/core/include/ecal/ecal_payload.h]: It's historical the prefix for a class. We will not follow these "rules" anymore but we keep the style for the core.

File: ecal/core/include/ecal/ecal_payload.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+    CPayload(const CPayload &) = delete;
+    CPayload(CPayload &&) noexcept = default;
+
+    CPayload& operator=(const CPayload &) = delete;
+    CPayload& operator=(CPayload &&) noexcept = default;
+
+    // the provisioned memory is uninitialized ->
+    // perform a full write operation
+    virtual bool WriteComplete (void* buf_, size_t len_) = 0;
```
[the remark this answers is not in the document]
Comment (rex-schilasky) [reply ecal/core/include/ecal/ecal_payload.h]: The core (caller) will just log an error.

[the remark this answers is not in the document]
Comment (rex-schilasky) [reply ecal/core/include/ecal/ecal_payload.h]: Renamed

**Pull request closed without merging**